snuba/datasets/sessions.py

```diff
@@ -0,0 +1,270 @@
+from snuba.query.project_extension import ProjectExtension, ProjectExtensionProcessor
+from snuba.query.query import Query
+from snuba.query.parsing import ParsingContext
+
+
+WRITE_LOCAL_TABLE_NAME = "sessions_raw_local"
+WRITE_DIST_TABLE_NAME = "sessions_raw_dist"
+READ_LOCAL_TABLE_NAME = "sessions_hourly_local"
+READ_DIST_TABLE_NAME = "sessions_hourly_dist"
+READ_LOCAL_MV_NAME = "sessions_hourly_mv_local"
+READ_DIST_MV_NAME = "sessions_hourly_mv_dist"
+
+STATUS_MAPPING = {
+    "ok": 0,
+    "exited": 1,
+    "crashed": 2,
+    "abnormal": 3,
+    "degraded": 4,
+}
+REVERSE_STATUS_MAPPING = {v: k for (k, v) in STATUS_MAPPING.items()}
+NIL_UUID = "00000000-0000-0000-0000-000000000000"
+
+
+class SessionsProcessor(MessageProcessor):
+    def process_message(self, message, metadata=None) -> Optional[ProcessedMessage]:
+        action_type = ProcessorAction.INSERT
+        if message["duration"] is None:
+            duration = None
+        else:
+            duration = _collapse_uint32(int(message["duration"] * 1000))
+
+        # since duration is not nullable, the max duration means no duration
+        if duration is None:
+            duration = MAX_UINT32
+
+        processed = {
+            "session_id": str(uuid.UUID(message["session_id"])),
+            "distinct_id": str(uuid.UUID(message.get("distinct_id") or NIL_UUID)),
+            "seq": message["seq"],
+            "deleted": 0,
+            "org_id": message["org_id"],
+            "project_id": message["project_id"],
+            "retention_days": message["retention_days"],
+            "duration": duration,
+            "status": STATUS_MAPPING[message["status"]],
+            "errors": _collapse_uint16(message["errors"]) or 0,
+            "timestamp": _ensure_valid_date(
+                datetime.utcfromtimestamp(message["timestamp"])
+            ),
+            "started": _ensure_valid_date(
+                datetime.utcfromtimestamp(message["started"])
+            ),
+            "release": message.get("release"),
+            "environment": message.get("environment"),
+        }
+        return ProcessedMessage(action=action_type, data=[processed])
+
+
+class SessionDataset(TimeSeriesDataset):
+    def __init__(self) -> None:
+        all_columns = ColumnSet(
+            [
+                ("session_id", UUID()),
+                ("distinct_id", UUID()),
+                ("seq", UInt(64)),
+                ("org_id", UInt(64)),
+                ("project_id", UInt(64)),
+                ("retention_days", UInt(16)),
+                ("duration", UInt(32)),
+                ("status", UInt(8)),
+                ("errors", UInt(16)),
+                # note here that deleted is really only an indicator so we can
+                # figure out which values can be discarded.  Deleting of values
+                # does not update the materializations.
+                ("deleted", UInt(8)),
+                ("timestamp", DateTime()),
+                ("started", DateTime()),
```

Comment:
We decided against calling `started` to mean `timestamp` on the protocol because it might mean that implementors accidentally misinterpret it.

Reply:
Then can we at least avoid `timestamp` and have `started` and `received` ? Timestamp is the column people generally use to query Snuba. Here it would make no sense. So if there are reasons for not renaming started to timestamp, let's change the name of received as well to reduce confusion.